Fixes to mixin forwarders

lrytz opened the pull request.

lrytz committed:
SD-210 don't generate invalid forwarders under -Xgen-mixin-forwarders

With -Xgen-mixin-forwarders the compiler eagerly creates mixin
forwarders for methods inherited from traits, even if the JVM method
resolution would pick the correct default method.

When generating a such a forwarder for a Java-defined default method,
the mixin forwarder invokes the default method directly using
`invokespecial` (for Scala-defined trait methods, the forwarder uses
the static `m$` method that is generated for every trait method).

An `invokespecial` call is only legal if the named interface is a
direct parent of the current class. If this is not the case, we don't
generate the mixin forwarder and emit a warning.

In the tests there's also an example where a mixin forwarder is
required for correctness, but cannot be added because the corresponding
`invokespecial` call would be invalid. In this case we emit an error.
This is similar to what we already do for other super calls to Java-
defined default methods. The difference is that the super call is not
written by the user but generated by the mixin forwarder. The solution
is the same: add the interface as a direct parent.

lrytz committed:
SD-143 error for super calls that cannot be implemented correctly

If a call super[T].m resolves to a method A.m where A is a class, but
not the direct superclass of the current class, there is no way to
emit an invocation of A.m: `invokespecial A.m` will resolve to B.m
where B is the superclass of the current class.

This commit adds an error message in this case.

Note that this is similar to an existing error message for qualified
super calls to a non-direct superclass:

    class A { def m = 1 }
    class B extends A { override def m = 2 }
    class C extends B { override def m = super[A].m }

Gives "error: A does not name a parent class of class C".

If the user means to call method m in the superclass, he can write an
unqualified `super.m`.

An similar error message is introduced if A is a Java-defined interface
(and m is a default method), and A is not a direct parent of the current
class. In this case `invokespecial A.m` is invalid bytecode. The
solution is to add A as a direct parent of the current class.

sjrd commented on src/compiler/scala/tools/nsc/transform/Mixin.scala:
```diff
@@ -243,11 +243,17 @@ abstract class Mixin extends InfoTransform with ast.TreeDSL {
               }
 
-              if (settings.XgenMixinForwarders) genForwarder()
+              if (settings.XgenMixinForwarders) genForwarder(required = false)
```
I think this will erroneously report a warning instead of an error if both of the following apply:
1. `-XgenMixinForwarders` is used, _and_
2. the forwarder is required for correctness.

I would expect an error to be raised every time a forwarder is required for correctness, regardless of the value of `settings.XgenMixinForwarders`.

lrytz replied:
good point, i have to change the order of tests

sjrd commented:
:+1: for the commit about JUnit. I think it's a much better experience to offer those by default, even though it does feel like burning library-specific things in the compiler.

retronym committed:
Disable stack hungry test of deprecated PagedSeq

lrytz committed:
Emit mixin forwarders for JUnit-annotated trait methods by default

JUnit 4 does not support default methods. For better user experience,
this commit makes the compiler generate mixin forwarders for inherited
trait methods that carry a JUnit annotation.

The -Yjunit-trait-methods-no-forwarders flag disables this behavior.

This supersedes the scala-js/scala-2.12-junit-mixin-plugin compiler
plugin.

lrytz commented:
will re-open

lrytz commented:
About the last commit: by default the compiler would now generate mixin forwarders whenever a class inherits a JUnit-annotated default method from a trait. For

```
trait T {
  @Test def t(): Unit = { ...}
}
class C extends T
```

we now get (by default):

```
public interface T {
    @Test
    default public static /* synthetic */ void t$(T $this) { $this.t(); }

    @Test
    default public void t() { }

    default public static void $init$(T $this) { }
}
public class C implements T {
    @Test
    @Override
    public void t() { T.t$(this); }
    public C() { T.$init$(this); }
}
```

This is certainly good for everyone using JUnit 4, which will ignore the default method and the static method (both annotated `@Test`) in T, and it will only execute `C.t`.

I did a quick test with the current master branch of JUnit 5 (https://github.com/junit-team/junit5). First of all there's a new `@Test` annotation in package `org.junit.jupiter.api`, but in some ways JUnit 5 depends on / supports JUnit 4, so the old `@org.junit.Test` might also be available and work, not sure. In any case, for the above source code, JUnit 5 executes the test `t()` in both cases, i.e., with and without a mixin forwarder. It doesn't (try to) run the static `t$` method.

So the current behavior (add forwarders for JUnit-annotated methods) seems to be fine also for JUnit 5, at least for its current state - not sure how things may change until their release.

Note that we don't emit forwarders when using the new `@org.junit.jupiter.api.Test`, which is fine: the annotation only exists in JUnit 5, which doesn't need the forwarder.


If we want to be more careful, instead of emitting the forwarders by default, we could provide a triple-choice command-line flag `-Yjunit-default-methods:<warn | gen-forwarders | nothing>` and set the default to `warn`.

adriaanm commented:
Sounds like the right approach to me as well. Small nit about the option choices if we do that: true|false|warn is easier to remember

lrytz commented:
> Sounds like the right approach to me as well

Which, to only warn by default?

lrytz commented:
Also, should it be a `-Y` option? Any thoughts on naming are welcome :)

adriaanm commented:
Yes to all the above :-) the main use case will be junit, which is done by
default, right? The only way to turn that off will be to explicitly pass
false to the option that I shall not name :-)
On Wed, Aug 31, 2016 at 14:36 Lukas Rytz <notifications@github.com> wrote:

> Also, should it be a -Y option? Any thoughts on naming are welcome :)
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/scala/scala/pull/5369#issuecomment-243750199>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAFjy89IdszXR8EIJNzjewvkcHYF_m_hks5qlXU7gaJpZM4Jw9pv>
> .
>

lrytz commented:
> Yes to all the above

OK, so "Yes, it should only warn, but not emit forwarders by default" and "yes, it should be a `-Y` option".

> the main use case will be junit

This is the only use case. we're not talking about emitting mixin forwarders in general, just for JUnit-annotated methods.

> which is done by default, right?

Well, if we set the default to warn, it will only warn.

adriaanm commented:
Ok then I misunderstood. I thought the default was to mix in forwarders for
test methods and warn for problems when we can't be sure to invoke the
right super method. So, yes to that. I interpreted 'true' to mean mix in
all forwarders and don't warn, and 'false' results in the minimal set of
forwarders, thus no junit test methods and no warnings.
On Wed, Aug 31, 2016 at 16:25 Lukas Rytz <notifications@github.com> wrote:

> Yes to all the above
>
> OK, so "Yes, it should only warn, but not emit forwarders by default" and
> "yes, it should be a -Y option".
>
> the main use case will be junit
>
> This is the only use case. we're not talking about emitting mixin
> forwarders in general, just for JUnit-annotated methods.
>
> which is done by default, right?
>
> Well, if we set the default to warn, it will only warn.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/scala/scala/pull/5369#issuecomment-243780868>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAFjy-9bwFmvWsdVNKtaKom8Elm_R5bUks5qlY7OgaJpZM4Jw9pv>
> .
>

lrytz commented:
OK. To recap, currently there are two independent flags:

  -  `-Xgen-mixin-forwarders` enables generating forwarders **always**. In case a forwarder cannot be generated, but the resulting bytecode is still correct, there's a warning
  - `-Yjunit-trait-methods-no-forwarders` disables generating forwarder for JUnit methods - by default they are generated.

Whenever a forwarder is required but cannot be generated, there's an error.

My proposal was to possibly extend the `-Y` option to a triple-choice (as explained above), and possibly set the default to `warn`. This is only about the JUnit case.

Another possibility, which is probably where you're getting at, would be to have one single option `-XYZmixin-forwarders` with the following choices:
  - `all` to generate all mixin forwarders, warn for those that cannot be emitted but are not required
  - `junit` to generate only forwarders for JUnit methods, error if one cannot be emitted
  - `junit-warn` to warn about JUnit methods in traits, but generate only forwarders required for correctness
  - `required-only` to only generate forwarders required for correctness

In all cases, if a required forwarder cannot be emitted, there's an error.

The default could be any of the above - to be decided. We might still go for `all` because of the performance impact (www.scala-lang.org/blog/2016/07/08/trait-method-performance.html), otherwise we could go for `junit` or `junit-warn`, I don't feel strongly about this.

adriaanm commented:
Yeah, I think a single option would be best. `-Xmixin-force-forwarders`?
  - `true`: all forwarders
  - `false`: minimal set of forwarders
  - `junit`: between `true` and `false` -- only ones we know are needed for JUnit 4

Thinking about it, how useful is `warn`? We know it's broken, and our users might not, so they won't know to turn on the option. Maybe we can just enable it implicitly, and disable when `-Xmixin-force-forwarders` is set explicitly (implying the user knows what's going on and doesn't need a warning).

lrytz commented:
> how useful is warn? We know it's broken

It's only broken for JUnit 4, when JUnit 5 arrives Scala 2.12 will still be around. OTOH, the forwarders should not hurt for JUnit 5 (as I tested above). So I think the `warn` level only makes sense if we put it as default. As an explicit option, I don't think it's too useful. The potential use case would be: somebody is using JUnit 4 and wants to make sure he's not writing any test methods in traits.

If we do have the warning, I'd prefer it to be an explicit option. The implicit may be a bit confusing.

lrytz commented:
The test failures in my commits are `PagedSeqTest.test_SI6615 failed: java.lang.StackOverflowError`. I rebased on current 2.12.x - is that expected? @szeiger

adriaanm commented:
Yeah apparently 1m of stack is not always enough. Jason has a commit in his
PR to just disable the test.
On Thu, Sep 1, 2016 at 09:10 Lukas Rytz <notifications@github.com> wrote:

> The test failures in my commits are PagedSeqTest.test_SI6615 failed:
> java.lang.StackOverflowError. I rebased on current 2.12.x - is that
> expected? @szeiger <https://github.com/szeiger>
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/scala/scala/pull/5369#issuecomment-243994941>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAFjyxZCGXkU01BvLLoX3LPVSonPDpu6ks5qlnpkgaJpZM4Jw9pv>
> .
>

lrytz committed:
Allow all settings to be helping

lrytz committed:
Allow per-choice help in ChoiceSetting

lrytz committed:
Add a -Xmixin-force-forwarders ChoiceSetting

lrytz commented:
OK, I think I'm done with the discussed changes, PTAL!

adriaanm commented on src/compiler/scala/tools/nsc/transform/Mixin.scala:
```diff
@@ -243,51 +243,59 @@ abstract class Mixin extends InfoTransform with ast.TreeDSL {
+                  JUnitAnnotations.exists(annot => annot.exists && member.hasAnnotation(annot))
+              }
+
+              if (existsCompetingMethod(clazz.baseClasses) || generateJUnitForwarder)
```
So, ignoring the annotations, `generateJUnitForwarder` is `false` iff `-Xmixin-force-forwarder:false`? When no value is supplied, when it's `true`, and when it's `junit` we get an error when annotations are present.

lrytz replied:
For JUnit methods, if `-Xmixin-force-forwarders` is either `true` or `junit` (which is the default, i.e., no value supplied), we generate the forwarder with `required = true`, i.e., there's an error if it cannot be generated.

If `-Xmixin-force-forwarder` is `false` we don't generate the JUnit forwarder.

adriaanm commented:
LGTM

lrytz commented:
pushed a small cleanup to the help texts of ChoiceSetting (https://github.com/scala/scala/pull/5369/commits/2f78fe11b1db18c44c748773507147de85dc7c48)
